src/ast2ram/ClauseTranslator.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2013, 2015, Oracle and/or its affiliates. All rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file ClauseTranslator.h
+ *
+ * Translator for clauses from AST to RAM
+ *
+ ***********************************************************************/
+
+#pragma once
+
+#include "ast2ram/AstToRamTranslator.h"
+#include "ast2ram/ValueIndex.h"
+#include <map>
+#include <vector>
+
+namespace souffle::ast {
+class Argument;
+class Clause;
+class Node;
+}  // namespace souffle::ast
+
+namespace souffle::ast::analysis {
+class AuxiliaryArityAnalsyis;
+}
+
+namespace souffle::ram {
+class Operation;
+class Condition;
+class RelationReference;
+}  // namespace souffle::ram
+
+namespace souffle::ast2ram {
+
+class ClauseTranslator {
+public:
+    ClauseTranslator(AstToRamTranslator& translator)
+            : translator(translator), auxArityAnalysis(translator.getAuxArityAnalysis()) {}
+
+    Own<ram::Statement> translateClause(
+            const ast::Clause& clause, const ast::Clause& originalClause, const int version = 0);
+
+protected:
+    AstToRamTranslator& translator;
+
+    // create value index
+    ValueIndex valueIndex;
+
+    // current nesting level
+    int level = 0;
+
+    virtual Own<ram::Operation> createOperation(const ast::Clause& clause);
+    virtual Own<ram::Condition> createCondition(const ast::Clause& originalClause);
+
+    /** translate RAM code for a constant value */
+    Own<ram::Operation> filterByConstraints(size_t level, const std::vector<ast::Argument*>& args,
+            Own<ram::Operation> op, bool constrainByFunctors = true);
+
+    const ast::analysis::AuxiliaryArityAnalysis* auxArityAnalysis;
+
+private:
+    // index nested variables and records
+    using arg_list = std::vector<ast::Argument*>;
+
+    std::vector<const ast::Argument*> generators;
+
+    // the order of processed operations
+    std::vector<const ast::Node*> op_nesting;
+
+    Own<ast::Clause> getReorderedClause(const ast::Clause& clause, const int version) const;
+
+    arg_list* getArgList(const ast::Node* curNode, std::map<const ast::Node*, Own<arg_list>>& nodeArgs) const;
```

Comment:
Does `nodeArgs` need to be a reference to a unique_ptr to a vector of raw Arguments pointers? A reference to the vector itself would make more sense.

Reply:
This is not a straightforward change. The reason `arg_list`s are being stored as unique pointer seems to be as a means of assigning unique ids (via their raw pointers) to each of them in the code (in `indexValues` and `createValueIndex`). I do agree it's odd, though. That section of the code should be refactored to not depend on that.

Reply:
I may change this to use the node pointer itself as the unique key. Since `nodeArg` uses `curNode` as a key to a unique value, and `argLevel` uses this value to map to a level, `curNode` can also be a unique key for `argLevel` directly.